pkg/controllers/leaderworkerset_controller.go

```diff
@@ -122,27 +122,61 @@ func (r *LeaderWorkerSetReconciler) Reconcile(ctx context.Context, req ctrl.Requ
 	return ctrl.Result{}, nil
 }
 
-func (r *LeaderWorkerSetReconciler) createHeadlessServiceIfNotExists(ctx context.Context, lws *leaderworkerset.LeaderWorkerSet) error {
+func (r *LeaderWorkerSetReconciler) reconcileHeadlessServices(ctx context.Context, lws *leaderworkerset.LeaderWorkerSet, replicas int32) error {
+
+	if lws.Spec.NetworkConfig == nil || lws.Spec.NetworkConfig.SubdomainPolicy == leaderworkerset.SubdomainShared {
+		if err := r.createHeadlessServiceIfNotExists(ctx, lws, lws.Name, map[string]string{leaderworkerset.SetNameLabelKey: lws.Name}); err != nil {
+			return err
+		}
+		return nil
+	}
+	// using replicas instead of lws.spec.replicas as needed to create services for burst replicas during maxSurge
+	for i := 0; i < int(replicas); i++ {
+		err := r.createHeadlessServiceIfNotExists(ctx, lws, fmt.Sprintf("%s-%s", lws.Name, strconv.Itoa(i)), map[string]string{leaderworkerset.GroupIndexLabelKey: strconv.Itoa(i)})
+		if err != nil {
+			return err
+		}
+	}
+
+	var headlessServiceList corev1.ServiceList
+	if err := r.List(ctx, &headlessServiceList, client.InNamespace(lws.Namespace)); err != nil {
+		return err
+	}
+
+	if len(headlessServiceList.Items) == int(replicas) {
+		return nil
+	}
+
+	// In case of scale down, need to delete the services of the replicas that will be deleted
+	for i := len(headlessServiceList.Items); i > int(replicas); i-- {
+		err := r.deleteHeadlessServiceIfExists(ctx, lws, fmt.Sprintf("%s-%s", lws.Name, strconv.Itoa(i-1)))
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *LeaderWorkerSetReconciler) createHeadlessServiceIfNotExists(ctx context.Context, lws *leaderworkerset.LeaderWorkerSet, serviceName string, serviceSelector map[string]string) error {
 	log := ctrl.LoggerFrom(ctx)
 	// If the headless service does not exist in the namespace, create it.
 	var headlessService corev1.Service
-	if err := r.Get(ctx, types.NamespacedName{Name: lws.Name, Namespace: lws.Namespace}, &headlessService); err != nil {
+	if err := r.Get(ctx, types.NamespacedName{Name: serviceName, Namespace: lws.Namespace}, &headlessService); err != nil {
 		if client.IgnoreNotFound(err) != nil {
 			return err
 		}
 		headlessService := corev1.Service{
 			ObjectMeta: metav1.ObjectMeta{
-				Name:      lws.Name,
+				Name:      serviceName,
 				Namespace: lws.Namespace,
 			},
 			Spec: corev1.ServiceSpec{
-				ClusterIP: "None", // defines service as headless
-				Selector: map[string]string{
-					leaderworkerset.SetNameLabelKey: lws.Name,
-				},
+				ClusterIP:                "None", // defines service as headless
+				Selector:                 serviceSelector,
 				PublishNotReadyAddresses: true,
 			},
 		}
+
 		// Set the controller owner reference for garbage collection and reconciliation.
 		if err := ctrl.SetControllerReference(lws, &headlessService, r.Scheme); err != nil {
```

Comment:
To manage deletions of headless services in an easier why, set the owner reference to the worker sts for the uniquePerReplica case. With this, you don't need to delete them explicitly because they will automatically be deleted when the worker sts is deleted. 

I would also be ok to not delete the headless service in the case we are switching from shared to uniquePerReplica, just leave a comment to that effect. This is because it makes the transition safer, and this headless service will be deleted eventually when the lws gets deleted anyway, so there is no resource leak.

Reply:
oh, right, this is a good point

Reply:
Changed